File: aion_custom_hr/public/js/employee_contract.js
frappe.ui.form.on('Employee', {
    refresh: function(frm) {
        // Ajouter le bouton "Create Contract" seulement si l'employé est sauvegardé
        if (!frm.is_new()) {
            frm.add_custom_button(__('Create Contract'), function() {
                create_employee_contract(frm);
            }, __('Actions'));
            
            // Optionnel: Ajouter un bouton pour voir les contrats existants
            frm.add_custom_button(__('View Contracts'), function() {
                view_employee_contracts(frm);
            }, __('Actions'));
            
            frm.add_custom_button(__('Job Opening'), function() {
                view_job_opening(frm);
            }, __('Actions'));
            
            // Ajouter le bouton pour voir l'historique des appraisals
            frm.add_custom_button(__('Voir Historique Appraisals'), function() {
                show_appraisal_history(frm);
            }, __('Appraisal'));
        }
    }
});

function view_job_opening(frm) {
    
        frappe.set_route("list", "Job Opening");
    
}

function create_employee_contract(frm) {
    if (!frm.doc.name) {
        frappe.msgprint(__('Please save the employee first'));
        return;
    }
    
    // Préparer les données du contrat avec les informations de l'employé
    let contract_data = {
        doctype: 'Contract',
        party_type: 'Employee',
        party_name: frm.doc.name,
        party_user: frm.doc.user_id,
        start_date: frm.doc.date_of_joining || frappe.datetime.get_today(),
        
        // Pré-remplir les termes du contrat avec des informations de base
        contract_terms: generate_employee_contract_terms(frm.doc),
        
        // Définir une date de fin par défaut (1 an après la date de début)
        end_date: frm.doc.contract_end_date || frappe.datetime.add_months(frm.doc.date_of_joining || frappe.datetime.get_today(), 12),
        
        // Références au document employé - laisser vide pour éviter les erreurs de validation
        document_type: '',
        document_name: '',
        
        // Statut initial
        status: 'Unsigned',
        requires_fulfilment: 1
    };
    
    // Créer un nouveau contrat avec les données pré-remplies
    frappe.new_doc('Contract', contract_data);
    
    frappe.show_alert({
        message: __('Contract draft created with employee information'),
        indicator: 'green'
    });
}

function generate_employee_contract_terms(employee_doc) {
    let terms = `
<h3>Employment Contract Terms</h3>

<p><strong>Employee Information:</strong></p>
<ul>
    <li>Employee Name: ${employee_doc.employee_name || 'N/A'}</li>
    <li>Employee ID: ${employee_doc.name || 'N/A'}</li>
    <li>Department: ${employee_doc.department || 'N/A'}</li>
    <li>Designation: ${employee_doc.designation || 'N/A'}</li>
    <li>Date of Joining: ${employee_doc.date_of_joining || 'N/A'}</li>
    <li>Employment Type: ${employee_doc.employment_type || 'N/A'}</li>
</ul>

<p><strong>Basic Terms:</strong></p>
<ul>
    <li>This employment contract establishes the terms and conditions of employment.</li>
    <li>The employee agrees to perform duties as assigned by the company.</li>
    <li>Salary and benefits as per company policy and applicable laws.</li>
    <li>Working hours and leave policy as per company handbook.</li>
</ul>

<p><strong>Additional Information:</strong></p>
<ul>
    <li>Branch: ${employee_doc.branch || 'N/A'}</li>
    <li>Company: ${employee_doc.company || 'N/A'}</li>
    <li>Reports To: ${employee_doc.reports_to || 'N/A'}</li>
</ul>

<p><em>Note: This is a template. Please customize the terms according to your organization's policies and legal requirements.</em></p>
    `;
    
    return terms;
}

function view_employee_contracts(frm) {
    if (!frm.doc.name) {
        frappe.msgprint(__('Please save the employee first'));
        return;
    }
    
    // Ouvrir la liste des contrats filtrée pour cet employé
    frappe.route_options = {
        "party_type": "Employee",
        "party_name": frm.doc.name
    };
    frappe.set_route("List", "Contract");
}

// Fonction pour ajouter des indicateurs visuels
frappe.ui.form.on('Employee', {
    onload: function(frm) {
        // Vérifier s'il y a des contrats existants pour cet employé
        if (!frm.is_new()) {
            check_existing_contracts(frm);
        }
    }
});

function check_existing_contracts(frm) {
    frappe.call({
        method: 'frappe.client.get_count',
        args: {
            doctype: 'Contract',
            filters: {
                party_type: 'Employee',
                party_name: frm.doc.name
            }
        },
        callback: function(r) {
            if (r.message > 0) {
                // Afficher un indicateur si des contrats existent
                frm.dashboard.add_indicator(__('Contracts: {0}', [r.message]), 
                    r.message > 0 ? 'blue' : 'grey');
                
                // Ajouter des informations supplémentaires
                get_contract_summary(frm);
            }
        }
    });
}

function get_contract_summary(frm) {
    frappe.call({
        method: 'frappe.client.get_list',
        args: {
            doctype: 'Contract',
            filters: {
                party_type: 'Employee',
                party_name: frm.doc.name
            },
            fields: ['name', 'status', 'start_date', 'end_date', 'is_signed'],
            limit: 5,
            order_by: 'creation desc'
        },
        callback: function(r) {
            if (r.message && r.message.length > 0) {
                let contract_info = '<div class="alert alert-info">';
                contract_info += '<strong>Recent Contracts:</strong><br>';
                
                r.message.forEach(function(contract) {
                    let status_indicator = contract.is_signed ? '✅' : '⏳';
                    contract_info += `${status_indicator} <a href="/app/contract/${contract.name}">${contract.name}</a> - ${contract.status}<br>`;
                });
                
                contract_info += '</div>';
                
                // Ajouter les informations au formulaire
                frm.set_df_property('employee_name', 'description', contract_info);
            }
        }
    });
}

function show_appraisal_history(frm) {
    // Récupérer les appraisals de l'employé
    frappe.call({
        method: 'frappe.client.get_list',
        args: {
            doctype: 'Appraisal',
            filters: {
                employee: frm.doc.name,
                docstatus: 1
            },
            fields: [
                'name',
                'employee',
                'employee_name',
                'total_score',
                'final_score',
                'modified',
                'appraisal_cycle'
            ],
            order_by: 'modified desc'
        },
        callback: function(r) {
            let all_appraisals = r.message || [];
            
            // Si l'employé a un manager, récupérer aussi ses appraisals
            if (frm.doc.reports_to) {
                frappe.call({
                    method: 'frappe.client.get_list',
                    args: {
                        doctype: 'Appraisal',
                        filters: {
                            employee: frm.doc.reports_to,
                            docstatus: 1
                        },
                        fields: [
                            'name',
                            'employee',
                            'employee_name',
                            'total_score',
                            'final_score',
                            'modified',
                            'appraisal_cycle'
                        ],
                        order_by: 'modified desc'
                    },
                    callback: function(r2) {
                        if (r2.message) {
                            all_appraisals = all_appraisals.concat(r2.message);
                        }
                        display_appraisals(all_appraisals, frm.doc.reports_to);
                    }
                });
            } else {
                display_appraisals(all_appraisals);
            }
        }
    });
}

function display_appraisals(appraisals, manager_id) {
    if (appraisals && appraisals.length > 0) {
        let message = '<div style="max-height: 400px; overflow-y: auto;">';
        message += '<table class="table table-bordered">';
        message += '<thead><tr>';
        message += '<th>Employé</th>';
        message += '<th>Date</th>';
        message += '<th>Référence</th>';
        message += '<th>Cycle</th>';
        message += '<th>Score Total</th>';
        message += '<th>Score Final</th>';
        message += '<th>Score Mensuel</th>';
        message += '</tr></thead>';
        message += '<tbody>';
        
        // Trier les appraisals par date
        appraisals.sort((a, b) => new Date(b.modified) - new Date(a.modified));
        
        appraisals.forEach(function(appraisal) {
            let monthly_score = (appraisal.total_score / 5) * 100;
            let date = frappe.datetime.str_to_user(appraisal.modified.split(' ')[0]);
            let is_manager = appraisal.employee === manager_id;
            
            message += `<tr ${is_manager ? 'style="background-color: #f8f9fa;"' : ''}>`;
            message += `<td>${appraisal.employee_name} ${is_manager ? '(Manager)' : ''}</td>`;
            message += `<td>${date}</td>`;
            message += `<td><a href="/app/appraisal/${appraisal.name}">${appraisal.name}</a></td>`;
            message += `<td>${appraisal.appraisal_cycle || ''}</td>`;
            message += `<td>${appraisal.total_score || 0}</td>`;
            message += `<td>${appraisal.final_score || 0}</td>`;
            message += `<td>${monthly_score.toFixed(2)}%</td>`;
            message += '</tr>';
        });
        
        message += '</tbody></table></div>';
        
        // Afficher le message avec les appraisals
        frappe.msgprint({
            title: __('Historique des Appraisals'),
            message: message,
            wide: true
        });
    } else {
        frappe.msgprint(__('Aucun appraisal trouvé.'));
    }
}

File: aion_custom_hr/api/employee_contract.py
import frappe
from frappe import _
from frappe.utils import getdate, add_months, today


@frappe.whitelist()
def create_employee_contract(employee_id, contract_type="Standard", duration_months=12):
    """
    Crée un contrat pour un employé avec des données pré-remplies
    """
    if not employee_id:
        frappe.throw(_("Employee ID is required"))
    
    # Récupérer les informations de l'employé
    employee = frappe.get_doc("Employee", employee_id)
    
    # Créer le contrat avec les données de l'employé
    contract = frappe.new_doc("Contract")
    
    # Informations de base
    contract.party_type = "Employee"
    contract.party_name = employee.name
    contract.party_user = employee.user_id
    
    # Dates
    contract.start_date = employee.date_of_joining or today()
    contract.end_date = add_months(contract.start_date, duration_months)
    
    # Termes du contrat
    contract.contract_terms = generate_contract_terms(employee, contract_type)
    
    # Références
    contract.document_type = ""
    contract.document_name = ""
    
    # Statut initial
    contract.status = "Unsigned"
    contract.requires_fulfilment = 1
    
    # Ajouter des termes d'accomplissement de base
    contract.append("fulfilment_terms", {
        "requirement": "Complete employee onboarding process",
        "notes": "Employee must complete all required onboarding documentation"
    })
    
    contract.append("fulfilment_terms", {
        "requirement": "Provide required identification documents",
        "notes": "Valid ID, passport, and other legal documents"
    })
    
    contract.append("fulfilment_terms", {
        "requirement": "Complete probation period evaluation",
        "notes": f"Evaluation to be completed after probation period"
    })
    
    # Sauvegarder le contrat
    contract.insert()
    
    return {
        "contract_name": contract.name,
        "message": _("Contract created successfully for employee {0}").format(employee.employee_name)
    }


def generate_contract_terms(employee, contract_type="Standard"):
    """
    Génère les termes du contrat basés sur les informations de l'employé
    """
    
    terms_template = f"""
    <h2>Employment Contract - {contract_type}</h2>
    
    <h3>Employee Information</h3>
    <table style="width: 100%; border-collapse: collapse; margin: 10px 0;">
        <tr style="background-color: #f5f5f5;">
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Employee Name:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.employee_name or 'N/A'}</td>
        </tr>
        <tr>
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Employee ID:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.name or 'N/A'}</td>
        </tr>
        <tr style="background-color: #f5f5f5;">
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Department:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.department or 'N/A'}</td>
        </tr>
        <tr>
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Designation:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.designation or 'N/A'}</td>
        </tr>
        <tr style="background-color: #f5f5f5;">
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Date of Joining:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.date_of_joining or 'N/A'}</td>
        </tr>
        <tr>
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Employment Type:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.employment_type or 'N/A'}</td>
        </tr>
        <tr style="background-color: #f5f5f5;">
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Branch:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.branch or 'N/A'}</td>
        </tr>
        <tr>
            <td style="padding: 8px; border: 1px solid #ddd;"><strong>Company:</strong></td>
            <td style="padding: 8px; border: 1px solid #ddd;">{employee.company or 'N/A'}</td>
        </tr>
    </table>
    
    <h3>Terms and Conditions</h3>
    
    <h4>1. Employment Relationship</h4>
    <p>This contract establishes an employment relationship between <strong>{employee.company or 'The Company'}</strong> 
    and <strong>{employee.employee_name or 'The Employee'}</strong> for the position of 
    <strong>{employee.designation or 'Employee'}</strong> in the <strong>{employee.department or 'General'}</strong> department.</p>
    
    <h4>2. Duties and Responsibilities</h4>
    <ul>
        <li>Perform assigned duties diligently and professionally</li>
        <li>Adhere to company policies and procedures</li>
        <li>Maintain confidentiality of company information</li>
        <li>Report to: <strong>{employee.reports_to or 'Direct Supervisor'}</strong></li>
    </ul>
    
    <h4>3. Compensation and Benefits</h4>
    <ul>
        <li>Salary and benefits as per company policy</li>
        <li>Leave entitlements as per company handbook</li>
        <li>Performance-based incentives (if applicable)</li>
        <li>Medical and other benefits as per company policy</li>
    </ul>
    
    <h4>4. Working Hours</h4>
    <ul>
        <li>Standard working hours as per company policy</li>
        <li>Overtime compensation as per applicable labor laws</li>
        <li>Flexible working arrangements (if applicable)</li>
    </ul>
    
    <h4>5. Probation Period</h4>
    <p>The employee will be on probation for the initial period as per company policy. 
    Performance will be evaluated, and confirmation will be subject to satisfactory completion.</p>
    
    <h4>6. Termination</h4>
    <ul>
        <li>Either party may terminate this contract with appropriate notice</li>
        <li>Notice period as per company policy and applicable laws</li>
        <li>Termination procedures as outlined in employee handbook</li>
    </ul>
    
    <h4>7. Confidentiality and Non-Disclosure</h4>
    <p>The employee agrees to maintain strict confidentiality regarding company information, 
    trade secrets, and proprietary data both during and after employment.</p>
    
    <h4>8. Compliance</h4>
    <p>This contract is subject to applicable labor laws and regulations. 
    Any disputes will be resolved as per company policy and legal procedures.</p>
    
    <p><em><strong>Note:</strong> This is a template contract. Please review and customize 
    according to your organization's specific policies, legal requirements, and local labor laws.</em></p>
    """
    
    return terms_template


@frappe.whitelist()
def get_employee_contract_summary(employee_id):
    """
    Récupère un résumé des contrats pour un employé
    """
    if not employee_id:
        return []
    
    contracts = frappe.get_all("Contract", 
        filters={
            "party_type": "Employee",
            "party_name": employee_id
        },
        fields=[
            "name", "status", "start_date", "end_date", 
            "is_signed", "signed_on", "fulfilment_status"
        ],
        order_by="creation desc"
    )
    
    return contracts


@frappe.whitelist()
def check_contract_requirements(employee_id):
    """
    Vérifie si un employé a besoin d'un nouveau contrat
    """
    if not employee_id:
        return {"needs_contract": False, "message": "No employee specified"}
    
    # Vérifier les contrats existants
    active_contracts = frappe.get_all("Contract",
        filters={
            "party_type": "Employee",
            "party_name": employee_id,
            "status": "Active"
        }
    )
    
    if not active_contracts:
        return {
            "needs_contract": True,
            "message": "No active contract found for this employee"
        }
    
    # Vérifier les contrats qui expirent bientôt
    expiring_soon = frappe.get_all("Contract",
        filters={
            "party_type": "Employee", 
            "party_name": employee_id,
            "status": "Active",
            "end_date": ["<=", add_months(today(), 2)]  # Expire dans 2 mois
        }
    )
    
    if expiring_soon:
        return {
            "needs_contract": True,
            "message": "Contract expiring soon - renewal may be required"
        }
    
    return {
        "needs_contract": False,
        "message": "Employee has active contract"
    }
